### PR TITLE
feat: Support exporting with dark mode

### PR DESCRIPTION
**Pull request by h7y**
Works with both SVG and PNG. 

### Example PNG
Example PNG

### Example SVG
Example SVG

Resolves:
#2026

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -107,3 +107,5 @@ export const MODES = {
   ZEN: "zenMode",
   GRID: "gridMode",
 };
+
```
Comment (lipis) [inline src/constants.ts:110]:
```suggestion

// As found in theme.scss
```

**Comment by dwelle** [comment]
Hey @h7y, thanks for the PR! As discussed in the issue, the darkMode export should be a separate toggle in the export dialog, and not be connected to current appearance state. See https://github.com/excalidraw/excalidraw/issues/2026#issuecomment-760910210

**Comment by h7y** [comment]
Hey @dwelle, I was under the assumption that this comment by @lipis was the final decision https://github.com/excalidraw/excalidraw/issues/2026#issuecomment-769689915 Should I go ahead and add a toggle?

**Comment by dwelle** [comment]
Ok, I completely missed (or misread) that last exchange, so the joke's on me. That said, I do not agree with that conclusion, we should get back to the drawing board I guess :).

**Comment by lipis** [comment]
It wasn't a conclusion or anything.. I was a bit sarcastic, but the written text is sometimes it's not obvious :) My bad..

Having said that though.. I didn't want to introduce more options in different places.. that's all.

**Comment by dwelle** [comment]
> I didn't want to introduce more options in different places.. that's all.

Yea, no one wants that. Sometimes, it's necessary though :(.

**Comment by h7y** [comment]
So I'll make the change?

**Comment by h7y** [comment]
> So I'll make the change?

@lipis @dwelle Should I?

**Comment by dwelle** [comment]
@h7y sorry, we're not ignoring you. The issue [doesn't seem settled yet](https://github.com/excalidraw/excalidraw/issues/2026), but if we're gonna run in circles we'll make a decision soon.

**Comment by dwelle** [comment]
@h7y it seems the issue discussion is indeed going in circles, so let's stick to the least controversial proposal which is to have a separate toggle in the export dialog, with separate state.

I'd put it in (1) position, but if it ends up covering the preview, you can go with (2).

image

**Comment by h7y** [comment]
It doesn't seem to cover the preview. I added it in (1).

**Comment by dwelle** [comment]
I was thinking about using the same darkMode toggle icon we use for the canvas darkMode, without the text (which we'd put in a tooltip)? /cc @lipis

**Comment by h7y** [comment]
Oooh makes sense! I'll do that.

File: src/actions/actionExport.tsx [hunk 2]
```diff
@@ -214,20 +215,20 @@ export const actionExportWithDarkMode = register({
...
       }}
     >
-      <input
-        type="checkbox"
-        checked={appState.exportWithDarkMode}
-        onChange={(event) => updateData(event.target.checked)}
-      />{" "}
-      {t("labels.exportWithDarkMode")}
-    </label>
+      <DarkModeToggle
```
Comment (dwelle) [inline src/actions/actionExport.tsx:226]: Let's make the `DarkModeToggle` component take `props.title` so we can supply a different tooltip for the export.

Comment (h7y) [reply src/actions/actionExport.tsx:226]: What should we pass as title? "Export with dark mode" or "Export with light mode" depending on state?

Comment (dwelle) [reply src/actions/actionExport.tsx:226]: Maybe "Toggle export color scheme" for both is enough?

Comment (h7y) [reply src/actions/actionExport.tsx:226]: For some reason, when passing the title this way https://github.com/excalidraw/excalidraw/pull/3046/commits/da8a587faf1e8083cbb1d3f80c8b16fdd2a7d6b0 somehow the other instance where we use this component to change background overrides title.

Comment (h7y) [reply src/actions/actionExport.tsx:226]: Should work now!

**Comment by dwelle** [comment]
I've noticed the dark mode export scheme is enabled only if the `export with background` is enabled. We should investigate why.

**Comment by h7y** [comment]
Also, just found that `context.filter` is not supported by Safari. I came across a [polyfill](https://github.com/davidenke/context-filter-polyfill). Should we use that or I can extract the things we need from the polyfill library and put it in the PR to get it working on Safari.

**Comment by dwelle** [comment]
> Should we use that or I can extract the things we need from the polyfill library and put it in the PR to get it working on Safari.

I'd first ship without the support for Safari then, and investigate the polyfill in a later PR.

It seems we can feature detect `filter` support via: 

```
'filter' in document.createElement('canvas').getContext('2d')
```

We should hide the toggle button altogether if not supported.

**Review by lipis — CHANGES_REQUESTED** [review]
If you uncheck the `With background`, the toggle does nothing..

**Comment by h7y** [comment]
@dwelle the package build is failing due to an `ERROR in ../../css/variables.module.scss 1:0`. any clue on what's happening?

**Comment by dwelle** [comment]
I forgot, we must be importing `constants.ts` into files that do not have the css loader (such as utils). We can add the loader, or we can revert the `aae90b9` commit for now.

**Comment by h7y** [comment]
> 
> 
> I forgot, we must be importing `constants.ts` in packages that do not use the css loader (such as utils). We can add the loader, or we can revert the `aae90b9` commit for now.

Added the loader!

**Review by lipis — APPROVED** [review]
Looking good

File: src/packages/utils/package.json [hunk 3]
```diff
@@ -44,7 +44,10 @@
     "babel-loader": "8.2.2",
     "babel-plugin-transform-class-properties": "6.24.1",
     "cross-env": "7.0.3",
+    "css-loader": "5.0.2",
     "file-loader": "6.2.0",
+    "mini-css-extract-plugin": "1.3.8",
+    "sass-loader": "11.0.1",
```
Comment (lipis) [inline src/packages/utils/package.json:50]: do we need all those?

Comment (h7y) [reply src/packages/utils/package.json:50]: I took these from excalidraw package.

Comment (dwelle) [reply src/packages/utils/package.json:50]: technically, for this feature we only need `sass-loader`, but `css-loader` makes sense. No clue what `mini-css-extract-plugin` is for, would need to ask @ad1992. But I guess we can update it later.

Comment (h7y) [reply src/packages/utils/package.json:50]: I'll try removing it and check if build passes?

Comment (h7y) [reply src/packages/utils/package.json:49]: Even in `excalidraw` package `mini-css-extract-plugin` is not found anywhere in webpack config.

Comment (ad1992) [reply src/packages/utils/package.json:49]: `mini-css-extract-plugin` is not needed here. you can remove it, it basicially to extract CSS to separate file which is not needed here.

Comment (h7y) [reply src/packages/utils/package.json:49]: Got it! Removed it :)

**Comment by dwelle** [comment]
Ok, let's do this. Thanks @h7y for the work! :heart:

**Comment by h7y** [comment]
Woo hoo 🥳 Thank you @dwelle & @lipis! You guys rock ❤️

**Comment by xixixao** [comment]
The focus ring overlaps the exported image (this doesn't look very nice and is distracting), and the toggle is the first focused button (not sure whether that is a problem or not) - perhaps the problem is that the focus ring is immediately visible.

Screen Shot 2021-03-01 at 10 04 58

**Comment by johtso** [comment]
Just a bit of feedback, as a user trying to find this feature I couldn't find the toggle and assumed it wasn't possible. Only after finding this pull request did I look again and find it. It's possible a little too unobtrusive? I think the issue is that it isn't where any of the other toggles are, and the other toggles are also large checkboxes so you don't expect to be looking for a small icon.

image

**Comment by PissedCapslock** [comment]
Not sure if this is a known issue, but this toggle button doesn't appear on Safari (Version 16.1 (18614.2.9.1.12)). It does work on Chrome though, so a quick browser switch is an easy workaround.

**Comment by dwelle** [comment]
@PissedCapslock yes. Sadly, Safari doesn't support filters on `<canvas>` which we use for exporting in dark mode at the moment.

**Pull request merged**